common/src/main/java/io/pravega/common/io/serialization/RevisionDataOutputStream.java

```diff
@@ -0,0 +1,279 @@
+            return 8;
+        } else if (value > 0x3FFF) {
+            return 4;
+        } else if (value > 0x3F) {
+            return 2;
+        } else {
+            return 1;
+        }
+    }
+
+    @Override
+    public void writeCompactLong(long value) throws IOException {
+        if (value < COMPACT_LONG_MIN || value >= COMPACT_LONG_MAX) {
+            throw new IllegalArgumentException("writeCompactLong can only serialize non-negative longs up to 2^62.");
+        } else if (value > 0x3FFF_FFFF) {
+            // All 8 bytes
+            writeInt((int) (value >>> 32 | 0xC000_0000));
+            writeInt((int) value);
+        } else if (value > 0x3FFF) {
+            // Only 4 bytes.
+            writeInt((int) (value | 0x8000_0000));
+        } else if (value > 0x3F) {
+            // Only 2 bytes.
+            this.out.write((int) (value >>> 8 | 0x40 & 0xFF));
+            this.out.write((int) (value & 0xFF));
+        } else {
+            // 1 byte.
+            this.out.write((byte) value);
+        }
+    }
+
+    @Override
+    public int getCompactIntLength(int value) {
+        if (value < COMPACT_INT_MIN || value >= COMPACT_INT_MAX) {
+            throw new IllegalArgumentException("writeCompactInt can only serialize non-negative longs up to 2^30.");
+        } else if (value > 0x3F_FFFF) {
+            return 4;
+        } else if (value > 0x3FFF) {
+            return 3;
+        } else if (value > 0x3F) {
+            return 2;
+        } else {
+            return 1;
+        }
+    }
+
+    @Override
+    public void writeCompactInt(int value) throws IOException {
+        if (value < COMPACT_INT_MIN || value >= COMPACT_INT_MAX) {
+            throw new IllegalArgumentException("writeCompactInt can only serialize non-negative longs up to 2^30.");
+        } else if (value > 0x3F_FFFF) {
+            // All 4 bytes
+            this.out.write(value >>> 24 | 0xC0 & 0xFF);
+            this.out.write(value >>> 16 & 0xFF);
+            this.out.write(value >>> 8 & 0xFF);
+        } else if (value > 0x3FFF) {
+            // 3 bytes.
+            this.out.write(value >>> 16 | 0x80 & 0xFF);
+            this.out.write(value >>> 8 & 0xFF);
+        } else if (value > 0x3F) {
+            // 2 Bytes.
+            this.out.write(value >>> 8 | 0x40 & 0xFF);
+        }
+
+        // Last byte is always the same.
+        this.out.write(value & 0xFF);
+    }
+
+    @Override
+    public void writeUUID(UUID uuid) throws IOException {
+        writeLong(uuid.getMostSignificantBits());
+        writeLong(uuid.getLeastSignificantBits());
+    }
+
+    @Override
+    public <T> void writeCollection(Collection<T> collection, ElementSerializer<T> elementSerializer) throws IOException {
+        if (collection == null) {
+            writeCompactInt(0);
+            return;
+        }
+
+        writeCompactInt(collection.size());
+        for (T e : collection) {
+            elementSerializer.accept(this, e);
```

Comment:
we should be writing the  length of serialized element in the collection. 
element serializer may not write the length of entire element upfront.. and different revisions can be serialized and deserialized whereby length has to be specified somewhere

Reply:
See my answer above.